webrender/src/clip_stack.rs

```diff
@@ -0,0 +1,178 @@
+use internal_types::DeviceRect;
+use prim_store::{ClipData, GpuBlock32, PrimitiveClipSource, PrimitiveStore};
+use prim_store::{CLIP_DATA_GPU_SIZE, MASK_DATA_GPU_SIZE};
+use tiling::StackingContextIndex;
+use util::TransformedRect;
+use webrender_traits::{AuxiliaryLists, ImageMask};
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash)]
+pub struct ClipAddressRange {
+    pub start: GpuStoreAddress, // start GPU address
+    pub count: u32, // number of items, not bytes
+}
+
+type ImageMaskIndex = u16;
+
+#[derive(Debug, Clone, Copy, Eq, PartialEq, Hash)]
+pub struct MaskCacheKey {
+    pub layer_id: StackingContextIndex,
+    pub clip_range: ClipAddressRange,
+    pub image: Option<GpuStoreAddress>,
+}
+
+impl MaskCacheKey {
+    pub fn empty(layer_id: StackingContextIndex) -> MaskCacheKey {
+        MaskCacheKey {
+            layer_id: layer_id,
+            clip_range: ClipAddressRange {
+                start: GpuStoreAddress(0),
+                count: 0,
+            },
+            image: None,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct MaskCacheInfo {
+    pub key: MaskCacheKey,
+    // this is needed to update the ImageMaskData after the
+    // ResourceCache allocates/load the actual data
+    // will be simplified after the TextureCache upgrade
+    pub image: Option<ImageMask>,
+    pub device_rect: DeviceRect,
+    // Vec<layer transforms>
+    // Vec<layer mask images>
+}
+
+struct LayerInfo {
+    world_transform: Matrix4D<f32>,
+    _transformed_rect: TransformedRect,
+    local_rect: Rect<f32>,
+    parent_id: Option<StackingContextIndex>,
+}
+
+pub struct ClipRegionStack {
+    layers: HashMap<StackingContextIndex, LayerInfo>,
+    image_masks: Vec<ImageMask>,
+    current_layer_id: Option<StackingContextIndex>,
+    device_pixel_ratio: f32,
+}
+
+impl ClipRegionStack {
+    pub fn new(device_pixel_ratio: f32) -> ClipRegionStack {
+        ClipRegionStack {
+            layers: HashMap::new(),
+            image_masks: Vec::new(),
+            current_layer_id: None,
+            device_pixel_ratio: device_pixel_ratio,
+        }
+    }
+
+    pub fn reset(&mut self) {
+        self.layers.clear();
+        self.image_masks.clear();
+        self.current_layer_id = None;
+    }
+
+    pub fn generate(&mut self,
+                    source: &PrimitiveClipSource,
+                    clip_store: &mut GpuStore<GpuBlock32>,
+                    aux_lists: &AuxiliaryLists)
+                    -> Option<MaskCacheInfo> {
+        let mut clip_key = MaskCacheKey::empty(match self.current_layer_id {
+            Some(lid) => lid,
+            None => return None,
+        });
+        let layer = &self.layers[&clip_key.layer_id];
+        let mut local_rect = Some(layer.local_rect);
+
+        let image = match source {
+            &PrimitiveClipSource::NoClip => None,
+            &PrimitiveClipSource::Complex(rect, radius) => {
+                let address = clip_store.alloc(CLIP_DATA_GPU_SIZE);
```

Comment:
Ideally we don't want to be calling alloc() on gpu stores during generate. The reason is that the gpu stores are not re-created during scrolling, so this will end up growing the gpu stores during scrolling. What most of the primitives do is a gpu store alloc when the prim is first added (if needed), and then just get a (mutable) reference to that address when they need to update that data. I think that would work OK in this case?